source/common/common/utility.h

```diff
@@ -636,13 +636,72 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// A TrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
+// child will contain a vector of size 1. A node with no children will
+// contain an empty vector.
+//
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class TrieEntry {
+public:
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+  const TrieEntry* operator[](uint8_t i) const {
```

Comment:
nit: brief comments on these methods, please. (In particular the behavior when "i" is not in the Trie).

Reply:
Done.